// @nestjs/terminus/dist/terminus-bootstrap.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
var TerminusBootstrapService_1;
"use strict";
const common_1 = require("@nestjs/common");
const terminus_constants_1 = require("./terminus.constants");
const core_1 = require("@nestjs/core");
const terminus_1 = require("@godaddy/terminus");
/**
 * Bootstraps the third party Terminus library with the
 * configured Module options
 */
let TerminusBootstrapService = TerminusBootstrapService_1 = class TerminusBootstrapService {
    /**
     * Initializes the service
     * @param options The terminus module options
     * @param refHost The application reference host of NestJS which contains the http server instance
     * @param terminus The terminus instance
     */
    constructor(options, terminus, refHost) {
        this.options = options;
        this.terminus = terminus;
        this.refHost = refHost;
        /**
         * The NestJS logger
         */
        this.logger = new common_1.Logger(TerminusBootstrapService_1.name, true);
    }
    /**
     * Executes the given health indicators and stores the caused
     * errors and results
     * @param healthIndicators The health indicators which should get executed
     */
    executeHealthIndicators(healthIndicators) {
        return __awaiter(this, void 0, void 0, function* () {
            const results = [];
            const errors = [];
            yield Promise.all(healthIndicators
                // Register all promises
                .map(healthIndicator => healthIndicator())
                .map((p) => p.catch((error) => {
                if (error instanceof terminus_1.HealthCheckError) {
                    errors.push(error.causes);
                }
                else {
                    throw error;
                }
            }))
                .map((p) => p.then((result) => result && results.push(result))));
            return { results, errors };
        });
    }
    /**
     * Logs an error message of terminus
     * @param message The log message
     * @param error The error which was thrown
     */
    logError(message, error) {
        if (error.causes) {
            const healthError = error;
            message = `${message} ${JSON.stringify(healthError.causes)}`;
        }
        this.logger.error(message, error.stack);
    }
    /**
     * Logs the health check registration to the logger
     * @param healthChecks The health check map to log
     */
    logHealthCheckRegister(healthChecks) {
        Object.keys(healthChecks).forEach(endpoint => this.logger.log(`Mapped {${endpoint}, GET} healthcheck route`, 'TerminusExplorer'));
    }
    getHealthCheckExecutor(endpoint) {
        return () => __awaiter(this, void 0, void 0, function* () {
            const { results, errors } = yield this.executeHealthIndicators(endpoint.healthIndicators);
            const info = (results || [])
                .concat(errors || [])
                .reduce((previous, current) => Object.assign(previous, current), {});
            if (errors.length) {
                throw new terminus_1.HealthCheckError('Healthcheck failed', info);
            }
            else {
                return info;
            }
        });
    }
    /**
     * Returns the health check map using the configured health
     * indicators
     */
    getHealthChecksMap() {
        return this.options.endpoints.reduce((healthChecks, endpoint) => {
            healthChecks[endpoint.url] = this.getHealthCheckExecutor(endpoint);
            return healthChecks;
        }, {});
    }
    /**
     * Bootstraps the third party terminus library with
     * the given module options
     */
    bootstrapTerminus() {
        const healthChecks = this.getHealthChecksMap();
        this.terminus(this.httpServer, {
            healthChecks,
            // Use the logger of the user
            // or by the default logger if is not defined
            logger: this.options.logger || this.logError.bind(this),
        });
        this.logHealthCheckRegister(healthChecks);
    }
    /**
     * Gets called when the application gets bootstrapped.
     */
    onApplicationBootstrap() {
        // httpServer for express, instance.server for fastify
        this.httpServer = this.refHost.httpAdapter.getHttpServer();
        this.bootstrapTerminus();
    }
};
TerminusBootstrapService = TerminusBootstrapService_1 = __decorate([
    common_1.Injectable(),
    __param(0, common_1.Inject(terminus_constants_1.TERMINUS_MODULE_OPTIONS)),
    __param(1, common_1.Inject(terminus_constants_1.TERMINUS_LIB)),
    __metadata("design:paramtypes", [Object, Function, core_1.HttpAdapterHost])
], TerminusBootstrapService);
exports.TerminusBootstrapService = TerminusBootstrapService;
